src/mrack/providers/beaker.py

```diff
@@ -0,0 +1,333 @@
+                    f"{prev_status} -> {status})"
+                )
+                prev_status = status
+
+            if STATUS_MAP.get(status) == STATUS_ACTIVE:
+                break
+            elif STATUS_MAP.get(status) == STATUS_PROVISIONING:
+                await asyncio.sleep(self.poll_sleep)
+            elif STATUS_MAP.get(status) in [STATUS_ERROR, STATUS_DELETED]:
+                logger.warning(
+                    f"Job {beaker_id} has errored with status "
+                    f"{status} with result {resource['result']}"
+                )
+                break
+            else:
+                logger.error(
+                    f"Job {beaker_id} has swithced to status "
+                    f"{status} with result {resource['result']}"
+                )
+                break
+
+        resource.update({"JobID": beaker_id, "hname": req_hname})
+        return resource
+
+    async def provision_hosts(self, hosts):
+        """Provision hosts based on list of host requirements.
+
+        Issues provisioning and waits for it succeed. Raises exception if any of
+        the servers was not successfully provisioned. If that happens it issues deletion
+        of all already provisioned resources.
+
+        Return list of information about provisioned servers.
+        """
+        started = datetime.now()
+
+        count = len(hosts)
+        logger.info(f"Issuing provisioning of {count} hosts")
+        create_res = []
+        for req in hosts:
+            resource = self.create_server(req)
+            create_res.append(resource)
+
+        create_resps = await asyncio.gather(*create_res)
+        logger.info("Provisioning issued")
+
+        logger.info("Waiting for all Beaker hosts to be available")
+        wait_res = []
+        for resp in create_resps:
+            resource = self.wait_till_provisioned(resp)
+            wait_res.append(resource)
+
+        server_results = await asyncio.gather(*wait_res)
+
+        provisioned = datetime.now()
+        provi_duration = provisioned - started
+
+        logger.info("All Beaker hosts reached provisioning final state (Reserved)")
+        logger.info(f"Provisioning duration: {provi_duration}")
+
+        hosts = [self.to_host(srv) for srv in server_results]
+        for host in hosts:
+            logger.info(host)
+
+        return hosts
+
+    async def delete_host(self, host):
+        """Delete provisioned hosts based on input from provision_hosts."""
+        logger.info(f"Deleting Beaker host from Job {host.id}")
+        return self.hub.taskactions.stop(
+            host.id, "cancel", "Job has been cancelled by mrack."
+        )
+
+    async def delete_hosts(self, hosts):
+        """Issue deletion of all servers based on previous results from provisioning."""
+        logger.info("Issuing Beaker deletion")
+
+        delete_bkr = []
+        for host in hosts:
+            aws = self.delete_host(host)
```

Comment:
nitpick: I would avoid `aws` just because we also use those letters to refer to "AWS®", making the output of `grep` a little noisy.

Reply:
agree that it is noisy and also bit confusing, what var name would you suggest? It is a shortcut for awaitables. Quite used in https://docs.python.org/3/library/asyncio-task.html